Adding Pi and Tajima's D calculations as Eidos Functions

npb596 opened the pull request:
Hi Ben!

I've used the calcWattersonTheta function in my work and figured it was worth putting in Pi and Tajima's D as well. Feel free to check my calculations and code quality, I've bitten as much as possible off of the calcWattersonTheta function. Aside from rechecking the code itself I have run simple neutral simulations and they seem to fall in line with expectations. For example, at the beginning Watterson's theta is much higher than Pi, presumably due to high numbers of new low-frequency mutations, and Tajima's D is consequently very negative. Then sometime around 10N - 20N ticks burn-in Watterson's theta and pi seem about even and Tajima's D fluctuates around 0. The simple combinationTwo equation is necessary for pi's calculation.

Cheers,
Nick

bhaller commented:
Wow, great!  I'll tag @philippmesser and @petrelharp here; guys, it'd be great if you could have a look at what Nick has done here, since you probably understand it, and I don't.  :->  Once I've got a thumbs-up from one or both of you, I'll process this PR.

I'll probably add these to SLiM itself, rather than using the PR here, since this seems generally useful and polished.  :->  Thanks a bunch!

npb596 commented:
Thanks to you! I wasn't sure how best to submit new functions other than here, that would be awesome to see them in SLiM itself! I will eagerly await Philipp and/or Peter's opinions.

bhaller commented:
> Thanks to you! I wasn't sure how best to submit new functions other than here, that would be awesome to see them in SLiM itself! I will eagerly await Philipp and/or Peter's opinions.

The way you did it is great; merging it into SLiM requires extra glue, which I will provide.  :->

petrelharp commented on functions/calcPiTheta.txt:
```diff
@@ -0,0 +1,60 @@
+	varCount = genomes.mutationFrequenciesInGenomes(species.mutations) * size(genomes);
+	// total count of sequences subtracted by count of variant sequences equals count of invariant sequences
+	invarCount = size(genomes) - varCount;
+	// count of pairwise differences per site (assuming a biallelic site) is the product of counts of both alleles, this is then summed for all sites 
```
this phrasing in the comment is confusing, since sites are not necessarily biallelic

npb596 replied:
Agreed though not sure my new wording is any better. I mention this a bit now in my bigger comment and within the docstrings. I also changed the comment within the script now so it states where I got the idea of just multiplying ref and var alleles, since I don't see that notation often.

petrelharp commented on functions/calcTajimaD.txt:
```diff
@@ -0,0 +1,65 @@
+	k = size(muts);
+	n = genomes.size();
+	a_1 = sum(1 / 1:(n - 1));
+	a_2 = sum(1 / 1:(n - 1 ^ 2));
```
```suggestion
	a_2 = sum(1 / 1:((n - 1) ^ 2));
```
... right?

npb596 replied:
There is an issue here, thanks! But I now believe this should be the correct one and I should have it in my fork now:

a_2 = sum(1 / (1:(n - 1)) ^ 2);

I'm using Durrett 2008 section 2.4.1 for the exact formula. 1 / (n - 1) is a vector, all values in this vector are squared before the division from 1. I decided to test it more closely this time around with a population of only 10 individuals (20 genomes). The output should be equivalent to this python code:

1/19**2 + 1/18**2 + 1/17**2 + 1/16**2 + 1/15**2 + 1/14**2 + 1/13**2 + 1/12**2 + 1/11**2 +1/10**2 + 1/9**2 + +1.8**2 + 1/7**2 + 1/6**2 + 1/5**2 + 1/4**2 + 1/3**2 + 1/2**2 + 1/1**2

which equals 1.5780382439130234 and this is what I get with the above a_2. Both my original and your suggestion deviated from this. I guess that's the trickiest part to code with the vector stuff but I ran the calcTajimaD outputting every variable with this same small population to recheck and I'm more confident it's all working now. I guess I should upload a small script actually using these functions.

petrelharp replied:
Looks good; however,  note that I get the same value with your code and my suggestion (and indeed they should be equivalent).

npb596 replied:
Ah, thanks for checking again. I'm trying now with two identical Tajima's D functions (except differing by your a_2 and mine) and only returning a_2, and indeed they are the same. I think a week ago I was trying to simplify it by running a function for both a_2's only and must've made some copy error. And I guess I got confused by the parentheses but I see how yours is basically just being more explicit that 1:(n - 1) is squared with the extra parentheses but the order of operations does work the same in both.

petrelharp commented on functions/calcTajimaD.txt:
```diff
@@ -0,0 +1,65 @@
+	// do the calculation
+	// Pi and Watterson's theta functions divide by sequence length so this must be undone in Tajima's D
+	// Sequence length is constant (i.e. no missing data or indels) so this can be applied equally over both metrics
+	diff = (calcPiTheta(genomes) - calcWattersonsTheta(genomes)) * (species.chromosome.lastPosition + 1);
```
Shouldn't `calcPiTheta( )` have `start`, `end`, and `muts` arguments passed in as well?

npb596 replied:
That makes sense but I don't know how best to do this, as those should be optional. I think I need to set up a condition if the user does enter that info? As it is now, it works fine genome-wide but entering starts and ends to Tajima's D does not change the pi and WT calculations so I do need to fix this somehow, or just disallow the extra options.

petrelharp replied:
Can't you just do
```suggestion
	diff = (calcPiTheta(genomes, muts, start, end) - calcWattersonsTheta(genomes, muts, start, end)) * (species.chromosome.lastPosition + 1);
```
... ?

npb596 replied:
Yep, that works, thanks! Another thing I'm not entirely sure what I thinking. I think I was assuming if I put muts, starts, and ends into the functions it would return errors when the user never actually defined those, and I figured I'd try it later but never got to it. I see that the kinds of conditions I mentioned would be necessary are already in the function (e.g. if (isNULL(muts)) muts = species.mutations;), so it's already handling those cases and defining the variables appropriately. Running it this way seems to work as expected.

petrelharp commented on functions/calcTajimaD.txt:
```diff
@@ -0,0 +1,65 @@
+	// Sequence length is constant (i.e. no missing data or indels) so this can be applied equally over both metrics
+	diff = (calcPiTheta(genomes) - calcWattersonsTheta(genomes)) * (species.chromosome.lastPosition + 1);
+	// calculate standard deviation of covariance of pi and Watterson's theta
+	k = size(muts);
```
Isn't it the case that `calcWattersonsTheta( )` is giving us `k/a_1`? It seems more transparent to either (a) calculate watterson's theta directly here like that, or (b) use `calcWattersonsTheta( )` to get the value of `k` here. (This is not obvious because of various possible definitions like "number of mutations" or "number of sites with a mutation", etc.)

npb596 replied:
k here should represent number of sites with a mutation. Not sure what to do here because I see how it's simpler to just calculate k/a_1 and divide by sites (as is done in calcWattersonsTheta) since that information is present already but as in the above comment I need to change this so intervals work properly over calcPiTheta and calcWattersonTheta.

petrelharp replied:
ah, I see - perhaps a comment?

npb596 replied:
Since I kept the above interval handling, I have now just left a comment that the variables given also define Watterson's theta.

petrelharp commented:
I can take a careful look. But, to do that I need a docstring, so it's unambigious what's being calculated?  I can tell what the code does, but it would be helpful if you wrote down what you'd like it to be calculating, so we can make sure that it's doing what you want.

For instance:
- what is the endpoint behavior of `start` and `end` is (both are inclusive; hopefully this agrees with other methods?)
- how is `muts` used (I think by ignoring all other mutations, i.e., calculating the value you'd get if those mutations simply did not exist/were silently not genotypable)
- the returned value of pi is "per site" (not "per segregating site")


Perhaps most confusingly: stacked mutations. The code is treating each distinct mutation as a distinctly genotypable locus; so suppose there is one position with two mutations that we'll call `A` and `B`. So, there's four possible genotypes: `-` (no mutations), `A`, `B`, and `AB`. One approach to calculating pi would be to treat these as four distinct alleles at the one locus, getting (except for a factor of `1/(N-1)` to account for sampling-without-replacement):
```
p- * (1 - p-) + pA * (1 - pA) + pB * (1 - pB) + pAB * (1 - pAB)
```
Instead, the approach taken here is to treat these as two separate loci, getting
```
2 * pA * (1 - pA) + 2 * pB * (1 - pB)
```
I think this is a good way to go - I agree with the code as written. However, it needs explaining; for instance, a surprising consequence is that it is possible for pi to be greater than 1. (There's a thread in slim-discuss I think where someone ran into this and it was very confusing.) 

I've looked fairly carefully comparing the implementation of Tajima's D here to the [one in tskit](https://github.com/tskit-dev/tskit/blob/d3c59baddfb9b8c3f72cdf7a7c2ce4dd616f15e2/python/tskit/trees.py#L8690). And I don't *think* they are the same - for instance, I think the first term in the covariance in the two implmentations differ by a factor of `a_1` (in your notation). It's entirely possible I messed that one up. Could you have a careful look?

bhaller commented:
Thanks @petrelharp!  So yes, would be great to have a docstring for these functions, @npb596.  It should be written mirroring the style of calls like `calcFST()`, `calcHeterozygosity()`, etc., in the manual (section 25.19.2), and state precisely what it is that you intend these functions to do, with attention to the considerations Peter has pointed out.  Of course please state something like "This function was written by Nick Bailey" at the end to credit yourself.  Sound good?  Thanks!  :->

petrelharp commented:
(Note: I'd also be happy to help with the docstring; but we need to nail down the questions above.)

npb596 commented:
Thanks for the look-through, @petrelharp!

For you main comment, I've updated my fork of this repo https://github.com/npb596/SLiM-Extras/tree/master/functions but have not pushed that here yet because I assume that will start up a new request and thread (I haven't pushed a lot of forks...). The above link has new docstrings as text files. Since I copied a lot of the code from calcWattersonsTheta I also copied a lot of the docstring from that function but I would appreciate a look there as well. I mainly added some references to papers for the calculations and otherwise I'm pretty sure it's still applicable word-for-word. When I first wrote these I really just wanted quick genome-wide whole-population calculations of standard pi and Tajima's D as a quick way to check when my simulations reached equilibrium. Otherwise, they aren't really crucial for any work I'm doing now. Once I wrote them I figured it may be nice to make them more "production" quality. I originally didn't have any bug checks or options to take gene intervals, so I took a lot of that from calcWattersonsTheta. I don't know how much of that is helpful as to intent. I guess it's also worth pointing out this would be an example of a command used (only p1.genomes is required):

print(calcPiTheta(p1.genomes, muts = sim.mutationsOfType(m1), start = 10000, end = 20000));

To answer the first of three points you mention. I actually didn't test start and end options until today so thank you for mentioning it. It works in calcPiTheta as in calcWattersonsTheta where you can define some interval in the simulated genome for the calculation, instead of the whole genome. There actually was a small bug in my calcPiTheta that didn't allow this to happen, a holdover from my initial tests. I've fixed that now (a change from species.mutations to muts within the pi calc) and it can be seen in the above link.

Regarding the second point, the above bug maybe caused some confusion on the use of muts because it wasn't being used in the calculation since I left species.mutations, which was largely equivalent, in there. It's used to specify a type of mutation, if desired.

Regarding the third point, yes the pi returned is per site (averaged). Another thing I decided on by matching it to calcWattersonsTheta.

I did not think about stacking at all, so that's helpful to know. I did consider the fact that the calculation assumes an infinite-sites model whereas a SLiM population can have multi-allelic sites, that was why I put the unclear comment on bi-allelic sites, so the calculation won't always be exact. I didn't want to bother with a multi-allelic implementation and this matches calcWattersonsTheta better anyways. The docstring there mentions this issue and I kept it in my docstrings, though also with citations for multi-allelic models to add to the point.

Lastly, I did look at your tskit link and I agree it's different. My code is similar to other implementations I've seen (https://github.com/cggh/scikit-allel/blob/master/allel/stats/diversity.py#L875) and I based it on a textbook https://services.math.duke.edu/~rtd/Gbook/Gbook.html (but still messed up the a2, thanks for checking yourself!). My notation also is based on those sources. Given that, I think the tskit one is flawed. It looks like tskit a and b are equivalent to c_1 and c_2 in my notation (it skips the steps of defining b_1 and b_2, which is fine). The problem is c_1 and c_2 are plugged into the final D equation where e_1 and e_2 should be. Per Durrett 2008 (pg. 66), 

e_1 = c_1 / a_1

e_2 = c1 / (a_1 ^ 2 + a_2)

I also was curious what the necessity of making h and g arrays is. It's probably something having to do with how the function is used that I'm not aware of. It stood out just cause I'm used to the version as in scikit-allel linked above. Hope that helps!

npb596 commented:
I took a second look at the tskit code and realised now I made a slight error in my description. e_2 is correctly implemented in tskit (b / (h**2 + g)) but the issue still stands for e_1, which should be (a / h) in the tskit notation. So it is just a difference by a factor of "a_1" or "h" like you said.

petrelharp commented:
Uh-oh: I'll have to take a close look. I got the tskit version looking at Tajima's paper, so it's possible there's a discrepancy between Tajima and Durrett...

npb596 commented:
Durrett is helpful for me in terms of concision but that makes sense to check up on Tajima's original paper too. They use the same variable names and it seems to me they still match up. If it helps, Tajima defines the variables in the following equations in his paper:

a_1 in equation 3
a_2 in equation 4
b_1 in equation 8
b_2 in equation 9
c_1 in equation 31
c_2 in equation 32
e_1 in equation 36
e_2 in equation 37
pi in equation 10 and 11 (10 being the one I've used)
Watterson's theta in equation 5
Tajima's D in equation 38

I think they all match up to how Durrett does it in his 2.4.1 section.

bhaller commented:
I've been 50% following along, while on vacation.  I saw the recent commits to this PR.  Is it now ready to merge (although I won't merge it, I'll integrate it in a different way), or is there more remaining to be done?  I'll be back from vacation soon and will catch up on this and other things.

npb596 commented:
I made a minor change to a comment just now but otherwise I have made all changes I intend to and as far as I know it's all functional. I figured I'd wait to see if @petrelharp caught anything else but I think I addressed everything he mentioned before.

bhaller commented:
I'll be back home in about 2 days and will work on this soon thereafter.  @petrelharp is this ready to go, from your perspective?  I'll probably have a question or two once I'm actually doing the work of merging it in.  Thanks folks!

bhaller commented:
Hi @npb596.  Looking at this now.  I see you wrote:

> // This code is provided in the public domain free of restriction, please credit SLiM-Extras (https://github.com/MesserLab/SLiM-Extras) and myself, Nick Bailey, if you use it

The plan is that this PR will not be merged; instead, these functions will go into the SLiM core in the same way the other `calc...()` functions are now, and will be documented in the SLiM manual.  I'm happy to say there that they were written by you, as you have put in your docstrings (and do you want your affiliations in the docstrings too?).  If you really want me to, I can also include your request to be credited by people who use them, but (a) that's a bit unusual for a contribution like this to an open-source project, to ask to be credited individually when people use a particular sub-component of the open-source project, and (b) probably people won't notice the request anyway, or won't actually credit you, since most people aren't that conscientious.  :->  So.  Please make the docstrings "final", knowing that the comments/credits in the code files themselves will not be seen by end users in general; make them the way you want them, and I will use them.  Thanks!

bhaller commented:
The other thing it'd be great to have for this is a unit test or three.  I don't have tests for the other `calc...()` functions, IIRC, because it wasn't clear to me how to test them in a rigorous and repeatable manner, which is an uncomfortable fact.  Testing stochastic simulations is hard.  But if you see a reasonable way to test these functions, it'd be great, and I could incorporate it into the merge.  Food for thought.

bhaller commented:
The third thing is the name of the function `calcPiTheta()`.  Is that a good name, do you both think?  Will it be clear to people how `calcPiTheta()`, `calcHeterozygosity()`, and `calcPairHeterozygosity()` relate to each other?  Perhaps it would be good to add a bit to the docstring that does a compare-contrast with one or both of those other methods, to make this more clear?  @philippmesser and I have already been talking about the confusion surrounding "pi" and "heterozygosity", and were planning to write a bit about it for the manual; so feel free to punt on that, and Philipp and I can do it on our end after this PR has been merged.  But at a minimum, let's make sure the name of the function is optimal; we can't change it once it's public.  Is everybody happy with it?

npb596 commented:
Hi @bhaller.

Looks like you saw I did change how I did crediting like you mentioned. I had it originally in the function text files just because I copied it off other text files like readMatrix.txt in there but I was uncertain about it after having it in the docstring too, seemed redundant at least, and yeah I don't seriously expect people to credit it. I'm satisfied with it now as some proof I can put in my records.

Regarding unit testing I can try some things this week, so I would need more time on that. I have indeed re-run the various tests I had before with simple neutral scenarios. That's mainly 1) seeing if they roughly neutral expectations (e.g. pi fluctuates around 4 * Ne * mu and Tajima's D fluctuates around 0), 2) trying different window sizes and mutation types to see that those functions are actually different from just using all mutations and whole genome and 3) occasionally outputting individual variables in tiny sims with so few individuals I could do math by hand. But I agree it's worth doing something more systematic. I could at least do a better job of recording the above stuff. So I could plot the neutral expectation by the sim values for (1) and for (3) maybe write a small script with a set seed and very small population and genome size, output full and it should be possible to calculate by hand. For the latter point maybe also output VCFs and see if that match up to other program calculations (e.g. VCFtools, PopGenome, Pegas etc.). Although different programs also have subtly different ways of doing the calculations but since there's no sequencing error or missing data in a simulation it shouldn't be too big of an issue. Whatever I do I'll make sure to upload a toy script for it.

Regarding the name, I used calcPi early in testing and switched to calcPiTheta to emphasize it's the estimate of theta, so it parallels the name Watterson's theta. But I honestly think calcPi would be sufficient for most people to get the idea immediately. It's called pi in VCFtools (https://vcftools.github.io/man_0112a.html#SITE%20FILTERING%20OPTIONS) and it uses the greek letter pi in many textbooks (I reference two relatively recent ones in the doc string). But I realize there are popular programs that call it stuff like "Nucleotide diversity" (https://rdrr.io/cran/pegas/man/nuc.div.html) so I guess that could be like calcNucleotideDiversity. To me personally that sounds like you could be talking about any measure of genetic diversity at all. Also it's not unheard of to refer to it as "Tajima's theta" (https://academic.oup.com/genetics/article/181/2/701/6062903?login=false) and the earliest paper I know describing the formula as used in the function is Tajima 1983. So I guess something like calcTajimasTheta is a nice direct parallel to calcWattersonsTheta. But anyways I still lean to something like calcPi. Whatever name is used, I loaded the doc string with citations so it should be abundantly clear what formula is being used for anyone who doesn't want to trust the name alone.

So I know this is pretty long now but there is one final thing I want to ask about. I didn't really look at or test out calcHeterozygosity much until you mentioned it now. In my simple neutral simulations it is indeed quite close to pi (but not identical) and I assume it'd differ more breaking Hardy-Weinberg expectations, which may be worth doing in unit-testing. But since I looked at the source function I realized calcHeterozygosity averaged by a variable called "length" that is either the entire sequence or only the window defined if start and end are used. I had been using the species.chromosome.lastPosition + 1 idiom (like in Line 52 https://github.com/npb596/SLiM-Extras/blob/master/functions/calcPiTheta.txt) because this is used in calcWattersonsTheta. But this is a problem, right? The final averaging should use the length variable? Otherwise diversity in a window will be underestimated by using the entire chromosome length every time. Also "length" is defined in calcWattersonTheta if windows are specified, though seemingly never used for anything. If it's a problem there then I should also change it in my functions. As another little test today, I used an edited calcWattersonsTheta where I made this small change, and compared it to the original. I did this by making a chromosome of 100,000 base pairs (as always, totally simplistic neutral sim) and calculating these functions over the whole thing versus specifying a random 10,000 base pair window. Diversity should be about the same (maybe not at every point but averaging out over the whole sim) since the mutation process is random and the entire chrom behaves the same (e.g. no recombination hotspots or anything else fancy) I would think. But I could see with the current calcWattersonsTheta the 10,000 bp windows seemed consistently underestimated by 10x compared to the whole chromosome. Using the "length" instead of the whole species.chromosome.lastPosition made it so the windows values were on the same order of magnitude as whole chromosome. If I'm missing anything please let me know. I guess this is something else I should upload a script of at some point instead of just describing.

bhaller commented:
> Hi @bhaller.
> 
> Looks like you saw I did change how I did crediting like you mentioned... I'm satisfied with it now as some proof I can put in my records.

Yep, great.

> Regarding unit testing I can try some things this week, so I would need more time on that...

No problem; the ad hoc testing you've done thus far is good enough for now.  It'd be great to have a better testing regime for _all_ of these `calc...()` methods, but that's not necessarily your problem, and it doesn't need to delay this process.

> Regarding the name, I used calcPi early in testing and switched to calcPiTheta to emphasize it's the estimate of theta, so it parallels the name Watterson's theta... Whatever name is used, I loaded the doc string with citations so it should be abundantly clear what formula is being used for anyone who doesn't want to trust the name alone.

Great.  I'll wait for opinions from @petrelharp and @philippmesser on the name.  I don't know enough about these matters to have an opinion myself.

> So I know this is pretty long now but there is one final thing I want to ask about... But since I looked at the source function I realized calcHeterozygosity averaged by a variable called "length" that is either the entire sequence or only the window defined if start and end are used... But I could see with the current calcWattersonsTheta the 10,000 bp windows seemed consistently underestimated by 10x compared to the whole chromosome...

It sounds to me like you've found a bug in `calcWattersonsTheta()`!  This of course underlines the need for better testing of these functions, sigh.  I will fix that bug, I've made a note to myself.  Sounds like you should use `length` to factor in the window length, if it makes your results come out correctly rather than wrong; I'm not an expert on the math of these things, though, so I can't really confirm that definitively.

@petrelharp just mentioned to me that he remains puzzled by the discrepancy between what you're doing here and what he's doing in tskit, and he thinks one of them might be wrong.  He's going to look into that.  So until that question is resolved, we'll wait for him.

Thanks a bunch!  In the home stretch!  :->

bhaller commented:
> > So I know this is pretty long now but there is one final thing I want to ask about... But since I looked at the source function I realized calcHeterozygosity averaged by a variable called "length" that is either the entire sequence or only the window defined if start and end are used... But I could see with the current calcWattersonsTheta the 10,000 bp windows seemed consistently underestimated by 10x compared to the whole chromosome...
> 
> It sounds to me like you've found a bug in `calcWattersonsTheta()`! This of course underlines the need for better testing of these functions, sigh. I will fix that bug, I've made a note to myself. Sounds like you should use `length` to factor in the window length, if it makes your results come out correctly rather than wrong; I'm not an expert on the math of these things, though, so I can't really confirm that definitively.

The bug fix for `calcWattersonsTheta()` was trivial, and it was very clearly a bug, so I've just committed the fix for that.  So if you build the current head version of SLiM, you'll get the fixed `calcWattersonsTheta()` and can use that in your testing of your own function.  Hopefully that simplifies things on your end.  :->  Thanks again for that catch!

npb596 commented:
Thanks! I've put the same fix in my functions. Otherwise, I'm in no rush.

bhaller commented:
> > Regarding the name, I used calcPi early in testing and switched to calcPiTheta to emphasize it's the estimate of theta, so it parallels the name Watterson's theta... Whatever name is used, I loaded the doc string with citations so it should be abundantly clear what formula is being used for anyone who doesn't want to trust the name alone.
> 
> Great. I'll wait for opinions from @petrelharp and @philippmesser on the name. I don't know enough about these matters to have an opinion myself.

Philipp weighed in on Slack.  He says "my preference is clearly to call it calcPi".  His reasoning:

Theta is a population level parameter (Theta = 4Ne*mu). You can "estimate" Theta in some situations because it is approximately the value of pi.  That's what an "estimator" does. It then gets the "hat" symbol on top and you can call it Tajima's estimator of pi, for example.  Or you can estimate in a different way, for example by counting the number of segregating sites. Then you get another estimator, called Watterson's estimator.  But this function here is simply "measuring/calculating" pi.  So that's what it's name should be.

This makes sense to me, so @npb596, let's change it to `calcPi()` unless someone objects.  Thanks!

Other than that, I think we're waiting on @petrelharp's verdict, and the PR is otherwise ready for me to handle.  :->

petrelharp commented:
Well, nothing like Actually Comparing Things. I've gone and independently implemented (a) Wikipedia, (b) Tajima 1989, (c) Durrett 2008 (that's the 2nd ed), (d) tskit, (e) the tskit docstring, and (f) this implementation:
```
import numpy as np

def wikipedia(T, S, n):
    # squared denominator for Tajima's D
    # from https://en.wikipedia.org/wiki/Tajima%27s_D
    a1 = sum([1/i for i in range(1, n)])
    a2 = sum([1/i**2 for i in range(1, n)])
    b1 = (n+1) / 3 / (n-1)
    b2 = 2 * (n**2 + n + 3) / (9 * n * (n-1))
    c1 = b1 - 1/a1
    c2 = b2 - (n+2)/(a1 * n) + a2 / (a1**2)
    e1 = c1 / a1
    e2 = c2 / (a1**2 + a2)
    var = e1 * S + e2 * S * (S-1)
    return (T - S/a1) / np.sqrt(var)

def tajima(T, S, n):
    # squared denominator for Tajima's D
    # from https://doi.org/10.1093/genetics/123.3.585
    a1 = sum([1/i for i in range(1, n)])
    a2 = sum([1/i**2 for i in range(1, n)])
    b1 = (n + 1) / (3 * (n-1))
    b2 = 2 * (n**2 + n + 3) / (9 * n * (n-1))
    c1 = b1 - 1/a1
    c2 = b2 - (n+2) / (a1 * n) + a2 / (a1**2)
    e1 = c1 / a1
    e2 = c2 / (a1**2 + a2)
    var = e1 * S + e2 * S * (S-1)
    return (T - S/a1) / np.sqrt(var)

def tskit(T, S, n):
    # from the tskit implementation
    h = sum(1 / np.arange(1, n))
    g = sum(1 / np.arange(1, n) ** 2)
    a = (n + 1) / (3 * (n - 1) * h) - 1 / h**2
    b = (
        2 * (n**2 + n + 3) / (9 * n * (n - 1))
        - (n + 2) / (h * n)
        + g / h**2
    )
    var = a * S + (b / (h**2 + g)) * S * (S - 1)
    D = (T - S / h) / np.sqrt(var)
    return D

def tskit_doc(T, S, n):
    # from the tskit docstring
    h = sum(1 / np.arange(1, n))
    g = sum(1 / np.arange(1, n) ** 2)
    a = (n + 1) / (3 * (n - 1) * h) - 1 / h**2
    b = 2 * (n**2 + n + 3) / (9 * n * (n - 1)) - (n + 2) / (h * n) + g / h**2
    c = h**2 + g
    D = (T - S / h) / np.sqrt(a * S + (b / c) * S * (S - 1))
    return D

def durrett(T, S, n):
    # from Prob Models for Sequence Evolution, 2nd ed
    a1 = sum([1/i for i in range(1, n)])
    a2 = sum([1/i**2 for i in range(1, n)])
    b1 = (n+1) / (3 * (n-1))
    b2 = 2 * (n**2 + n + 3) / (9 * n * (n-1))
    c1 = b1 - 1/a1
    c2 = b2 - (n+2)/(a1 * n) + a2 / (a1**2)
    e1 = c1 / a1
    e2 = c2 / (a1**2 + a2)
    var = e1 * S + e2 * S * (S-1)
    return (T - S/a1) / np.sqrt(var)

def pr(T, S, n):
    # from https://github.com/MesserLab/SLiM-Extras/pull/28
    a_1 = sum([1 / i for i in range(1, n)])
    a_2 = sum([1 / i**2 for i in range(1, n)])
    b_1 = (n + 1) / (3 * (n - 1));
    b_2 = 2 * (n ** 2 + n + 3) / (9 * n * (n - 1));
    c_1 = b_1 - 1 / a_1;
    c_2 = b_2 - (n + 2) / (a_1 * n) + a_2 / a_1 ** 2;
    e_1 = c_1 / a_1;
    e_2 = c_2 / (a_1 ** 2 + a_2);
    var = e_1 * S + e_2 * S * (S - 1);
    D = (T - S / a_1) / np.sqrt(var)
    return D

for T, S, n in [(0.2, 15, 20), (0.1, 4, 5), (0.001, 9, 23)]:
    wiki = wikipedia(T, S, n)
    taj = tajima(T, S, n)
    tsk = tskit(T, S, n)
    tskd = tskit_doc(T, S, n)
    dur = durrett(T, S, n)
    npb = pr(T, S, n)
    print(f"({T}, {S}, {n}): ")
    print(f" wikipedia: {wiki}")
    print(f"    tajima: {taj}")
    print(f"     tskit: {tsk}")
    print(f"tskit docs: {tskd}")
    print(f"   Durrett: {dur}")
    print(f"    npb596: {npb}")
```
which produces
```
$ python test.py 
(0.2, 15, 20): 
 wikipedia: -3.515697435845563
    tajima: -3.515697435845563
     tskit: -3.515697435845563
tskit docs: -3.515697435845563
   Durrett: -3.515697435845563
    npb596: -3.515697435845563
(0.1, 4, 5): 
 wikipedia: -6.220982186871262
    tajima: -6.220982186871262
     tskit: -6.220982186871256
tskit docs: -6.220982186871256
   Durrett: -6.220982186871262
    npb596: -6.220982186871262
(0.001, 9, 23): 
 wikipedia: -3.311921355571379
    tajima: -3.311921355571379
     tskit: -3.3119213555713785
tskit docs: -3.3119213555713785
   Durrett: -3.311921355571379
    npb596: -3.311921355571379
```

So, I think we're all good - no discrepancies.

However, @npb596 - could you look back and double-check? In particular, see if you can identify what you thought was the error in the tskit implementation?

bhaller commented:
> Well, nothing like Actually Comparing Things. I've gone and independently implemented (a) Wikipedia, (b) Tajima 1989, (c) Durrett 2008 (that's the 2nd ed), (d) tskit, (e) the tskit docstring, and (f) this implementation...

Wow, that is a remarkably thorough check!  Thanks @petrelharp, looks great.  I will do the merging work as soon as this percolates to the top of my queue.  Thanks!  (BTW, chatting with @petrelharp about the name recently and he agrees, `calcPi()` is the correct name.)

npb596 commented:
Hi @petrelharp 

Thanks for checking again. It all looks good to me now. I believe I see where I misread the code, it is a lot of variables to keep track of! This is what I said before:

"Given that, I think the tskit one is flawed. It looks like tskit a and b are equivalent to c_1 and c_2 in my notation (it skips the steps of defining b_1 and b_2, which is fine). The problem is c_1 and c_2 are plugged into the final D equation where e_1 and e_2 should be."

and

"I took a second look at the tskit code and realised now I made a slight error in my description. e_2 is correctly implemented in tskit (b / (h**2 + g)) but the issue still stands for e_1, which should be (a / h) in the tskit notation. So it is just a difference by a factor of "a_1" or "h" like you said."

So I thought your a is my c_1 and your b is my c_2. I was right about b, but you do adjust it in the final formula to make it my e_2, which I realized in the second message. I now realize too that your a is actually my e_1. So it skips both defining b_1 and c_1.

```a = (n + 1) / (3 * (n - 1) * h) - 1 / h**2```

```e_1 = c_1 / a_1```

Since your h is my a_1, you are in dividing it out of c_1 by having it in the denominator of both factors in your a. So it's not an issue after all.

bhaller commented:
OK, ready to go, then.  I converted it to a draft PR so I don't accidentally press the merge button, lol.

bhaller commented:
FYI, I'm changing the name to `calcTajimasD()` to match the existing `calcWattersonsThete()`; maybe I should've named that `calcWattersonTheta()`, but here we are.  Consistency is the main thing.

bhaller commented:
In the docstrings, I should change "pi" and "theta" and such to the corresponding Greek letters, yes?  @petrelharp

petrelharp commented:
Great - thanks for the check, @npb596! It's very helpful to have more than one set of eyes on this kind of thing.

bhaller commented:
This is done; I've migrated the code and doc into SLiM itself.  I haven't tested it at all, but if it worked before, it should work now.  :->  @npb596 it'd be great if you had time to run your same tests with what are now SLiM built-in functions (note the name changes).  Thanks again for the contribution!  I love that SLiM is building up a set of these very useful pop-gen functions, and this is work I could not do myself.
